### PR TITLE
Better RC loss failsafe configuration options for BVLOS

### PR DESCRIPTION
**Pull request by MaEtUgR**
**Describe problem solved by this pull request**
Resolves https://github.com/PX4/PX4-Autopilot/pull/16864

Some Pilots in BVLOS (Beyond Visual Line Of Sight) scenarios want to be able to **fly foremost missions but presumably also loiter/offboard even when they lose the RC link that is present at boot**. To currently do that the only way is disabling the RC (Remote Control) loss reaction https://docs.px4.io/master/en/advanced_config/parameter_reference.html#NAV_RCL_ACT which is generally not safe to do if you ever fly relying on RC stick input because then the vehicle will ignore the link loss and just keep on using the last known value e.g. in stabilized mode.

I also realized that there are **no reactions in takeoff mode for link loss**. This could potentially result in a flyaway with an unreachable takeoff altitude with lost link.

**Describe your solution**
As discussed with @dagar and I think also @julianoes and others:
- I'm **not allowing to disable the RC loss action** anymore.
- Instead, I give the **option to specifically configure exceptions for certain modes such that the vehicle continues to fly even though a present RC is lost mid-flight**.
- Do link loss failsafe reactions in takeoff like in loiter/hold mode.

**Describe possible alternatives**
The reasoning is:
- By default, if a pilot connects an RC and flies whatever mode and has RC loss configured (default) he expects the failsafe to kick in. Typical use cases are flying missions or offboard with a safety pilot on-site watching the drone and expecting he can take over if something goes wrong at any point in time. In certain legislations this is even the mandatory workflow without a special permit.
- Allowing BVLOS operations in combination with RC for takeoff and landing for certain scenarios and products like @sanderux's.

Possible follow-ups:
- I'm aware that NAV_RCL_ACT can still be forced to disabled and that should be completely disallowed/ignored. It currently doesn't break backwards compatibility although I see no reason why anyone would disable the failsafe given the new options.
- A lot of the failsafe logic between autonomous modes are duplicated. I didn't go through unifying but it might be possible to simplify and make it more readable mostly in the interest of reducing error-proneness. E.g. it works correctly for mission but not some other mode that wasn't tested in the same corner case (e.g. takeoff).

**Test data / coverage**
I SITL tested the RC loss exception bits to act exactly like intended.
Failsafe in takeoff also works as expected.

The only thing I noticed which is independent of this pr is that the RC loss delay uses Hold mode to delay the vehicle's reaction but since goto (for some reason) is Hold mode the drone continues upon RC loss for the delay time towards the goto waypoint. I think we should really make goto a separate mode.

I noticed another thing: QGC seems to magically know about the parameter description that I just added without me recompiling with the adjusted reference xml. But there seems to be an off-by-one bug somewhere. The zeroth bit does not att anythign and the first bit adds one. I checked and other parameter bitfields are also defined zero based.
image

### CODE REVIEW

**Comment by sanderux** [comment]
Thanks @MaEtUgR looks good.
Testing will be a bit harder as I believe there are some issues with master on nora, need to check still

To clarify: when having a separate RC and GCS link, it is not uncommon that the GCS link has more range than the RC link. While GCS is still connected the pilot is still in control of the vehicle, rc loss should not be an issue when it is in an autonomous mode. A typical data link loss time would be around 1 minute or so when flying mapping missions. but going 1 minute without RC in a manual mode would be a big problem.

**Comment by MaEtUgR** [comment]
> To clarify: when having a separate RC and GCS link, it is not uncommon that the GCS link has more range than the RC link. While GCS is still connected the pilot is still in control of the vehicle, rc loss should not be an issue when it is in an autonomous mode. A typical data link loss time would be around 1 minute or so when flying mapping missions. but going 1 minute without RC in a manual mode would be a big problem.

Thanks for the insights. I've not had the pleasure of better data than RC link before but it's a valid use case and a matter of time until I'll see that too. With this pr it should be fully configurable e.g. you disable RC loss reaction for mission and/or for loiter depending on the use case. Datalink loss is handled independently and can also be disabled (default) in case the pilot really wants that. When all links are lost AND the mission is finished the vehicle always returns.

**Comment by MaEtUgR** [comment]
I fixed the style and wording on "Hold" vs "Loiter" by rebase.

File: src/modules/commander/commander_params.c [hunk 1]
```diff
@@ -833,7 +833,7 @@ PARAM_DEFINE_INT32(COM_FLIGHT_UUID, 0);
  *
  * @value 0 Hold
  * @value 1 Mission (if valid)
- * @group Mission
+ * @group Commander
```
Comment (julianoes) [inline src/modules/commander/commander_params.c:836]: What was your thought when changing that?

File: src/modules/commander/state_machine_helper.cpp [hunk 2]
```diff
@@ -436,15 +436,13 @@ bool set_nav_state(vehicle_status_s &status, actuator_armed_s &armed, commander_
...
 		   const offboard_loss_rc_actions_t offb_loss_rc_act,
 		   const position_nav_loss_actions_t posctl_nav_loss_act,
-		   const float param_com_rcl_act_t)
+		   const float param_com_rcl_act_t, const int param_com_rcl_except)
 {
 	const navigation_state_t nav_state_old = status.nav_state;
 
-	const bool data_link_loss_act_configured = data_link_loss_act > link_loss_actions_t::DISABLED;
-	const bool rc_loss_act_configured = rc_loss_act > link_loss_actions_t::DISABLED;
-	const bool rc_lost = rc_loss_act_configured && (status.rc_signal_lost);
```
Comment (julianoes) [inline src/modules/commander/state_machine_helper.cpp:445]: Nice, because that was ugly!

File: src/modules/commander/state_machine_helper.cpp [hunk 3]
```diff
@@ -539,27 +536,27 @@ bool set_nav_state(vehicle_status_s &status, actuator_armed_s &armed, commander_
...
 			enable_failsafe(status, old_failsafe, mavlink_log_pub, reason_no_datalink);
 			set_link_loss_nav_state(status, armed, status_flags, internal_state, rc_loss_act, 0);
 
-		} else if (status.rc_signal_lost && !rc_loss_act_configured
+		} else if (status.rc_signal_lost && (param_com_rcl_except & RCLossExceptionBits::RCL_EXCEPT_MISSION)
 			   && status.data_link_lost && !data_link_loss_act_configured
 			   && is_armed && !landed
 			   && mission_finished) {
-			// All links lost, all link loss reactions disabled -> return after mission
+			// All links lost, all link loss reactions disabled -> return after mission finished
```
Comment (julianoes) [inline src/modules/commander/state_machine_helper.cpp:559]: I think that makes sense.

File: src/modules/commander/state_machine_helper.cpp [hunk 4]
```diff
@@ -539,27 +536,27 @@ bool set_nav_state(vehicle_status_s &status, actuator_armed_s &armed, commander_
...
 			// Pilot disabled all reactions, finish mission but then return to avoid lost vehicle
 			enable_failsafe(status, old_failsafe, mavlink_log_pub, reason_no_datalink);
```
Comment (julianoes) [inline src/modules/commander/state_machine_helper.cpp:561]: Should there be a `reason_no_rc_and_no_datalink` here?

File: src/modules/commander/state_machine_helper.cpp [hunk 5]
```diff
@@ -580,25 +577,35 @@ bool set_nav_state(vehicle_status_s &status, actuator_armed_s &armed, commander_
...
-		} else if (status.rc_signal_lost) {
-			/* don't bother if RC is lost if datalink is connected */
-
-			/* this mode is ok, we don't need RC for LOITERing */
-			status.nav_state = vehicle_status_s::NAVIGATION_STATE_AUTO_LOITER;
+		} else if (status.rc_signal_lost && (param_com_rcl_except & RCLossExceptionBits::RCL_EXCEPT_LOITER)
+			   && status.data_link_lost && !data_link_loss_act_configured
+			   && is_armed && !landed) {
+			// All links lost, all link loss reactions disabled -> return
+			// Pilot disabled all reactions, return to avoid lost vehicle
```
Comment (julianoes) [inline src/modules/commander/state_machine_helper.cpp:604]: So that's now completely a "failsafe" for users but will be surprising for someone that actually wants that.

File: src/modules/commander/state_machine_helper.h [hunk 6]
```diff
@@ -115,6 +115,12 @@ enum class arm_disarm_reason_t {
 	UNIT_TEST = 13
 };
 
+enum RCLossExceptionBits {
```
Comment (julianoes) [inline src/modules/commander/state_machine_helper.h:118]: Is `enum class` not possible?

**Review by julianoes — COMMENTED** [review]
Nice work!

File: src/modules/commander/commander_params.c [hunk 7]
```diff
@@ -862,22 +864,37 @@ PARAM_DEFINE_INT32(NAV_DLL_ACT, 0);
...
 
+/**
+ * RC loss exceptions
+ *
+ * Specify modes in which RC loss is ignored and the failsafe action not triggered.
+ *
+ * @min 0
+ * @max 31
+ * @bit 0 Mission
+ * @bit 1 Loiter
```
[the remark this answers is not in the document]
Comment (MaEtUgR) [reply src/modules/commander/commander_params.c:887]: I wasn't sure what's the "real term" now loiter or hold. To me hold also makes more sense but I might have a multicopter biased perspective. I'm happy to use hold 🎉

File: src/modules/commander/commander_params.c (hunk 1 above)
Comment (MaEtUgR) [reply src/modules/commander/commander_params.c:836]: - They are among the commander parameters.
- They are used in commander and configure commander behavior.
- I thought the `NAV_` prefix just stayed because these parameters were moved from navigator to commander once but the names stayed backward compatible. This is in principle solvable with the parameter conversion.
- I'm assume some parameters like `COM_OBS_AVOID` that were added after moving parameters from navigator to commander even inherited the possibly wrong group from copy-pasting.

File: src/modules/commander/state_machine_helper.cpp (hunk 4 above)
Comment (MaEtUgR) [reply src/modules/commander/state_machine_helper.cpp:561]: There should. I would call it `reason_all_links_lost` but maybe your suggestion is more consistent.

File: src/modules/commander/state_machine_helper.h (hunk 6 above)
Comment (MaEtUgR) [reply src/modules/commander/state_machine_helper.h:118]: I'll check, I was honestly following existing patterns assuming it was already attempted there.

File: src/modules/commander/state_machine_helper.cpp (hunk 5 above)
Comment (MaEtUgR) [reply src/modules/commander/state_machine_helper.cpp:604]: So the use case would be to let the vehicle loiter without any link right? Is this more a simulation case or do you expect actual use cases e.g. switching from one link to another while loitering?

In any case the concern is valid. The reason I added it was because I expected a mission to command the vehicle to loiter for a certain time is enough but maybe it's not. Should we then remove the case and just react on losing either link and the user is able to disable the reaction (data link being disabled by default?).

I'd like this solution more if data link loss reaction was enabled by default but would only react on actual loss and not on never having any data link essentially making it a requirement to exist in the first place.

File: src/modules/commander/commander_params.c (hunk 1 above)
Comment (julianoes) [reply src/modules/commander/commander_params.c:836]: Ok fair.

File: src/modules/commander/commander_params.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (julianoes) [reply src/modules/commander/commander_params.c:887]: Just consistent with the other params.

File: src/modules/commander/state_machine_helper.cpp (hunk 4 above)
Comment (julianoes) [reply src/modules/commander/state_machine_helper.cpp:561]: Ah `reason_all_links_lost` works.

File: src/modules/commander/state_machine_helper.cpp (hunk 5 above)
Comment (julianoes) [reply src/modules/commander/state_machine_helper.cpp:604]: Ah, so this is only for loiter but not for mission?

File: src/modules/commander/state_machine_helper.cpp (hunk 4 above)
Comment (MaEtUgR) [reply src/modules/commander/state_machine_helper.cpp:561]: image

File: src/modules/commander/state_machine_helper.cpp (hunk 5 above)
Comment (MaEtUgR) [reply src/modules/commander/state_machine_helper.cpp:604]: I'll change it back for now and we'll improve e.g. with the default data link loss next iteration. After all this was not really the scope of this pr but just added because I had concerns when users might still accidentally disable all failsafes. Let's go in steps.

File: src/modules/commander/state_machine_helper.h (hunk 6 above)
Comment (MaEtUgR) [reply src/modules/commander/state_machine_helper.h:118]: Ah now I remember you'd have to implement the operators when you define your own class.
image
see also https://stackoverflow.com/questions/32578638/how-to-use-c11-enum-class-for-flags
So it's very cumbersome for bitfield definitions if you don't have a specific bitfield type to inherit from already. Not sure how much value that would add.

File: src/modules/commander/commander_params.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (MaEtUgR) [reply src/modules/commander/commander_params.c]: Changed using interactive rebase.
https://github.com/PX4/PX4-Autopilot/commit/79c9af5a79c546f3b1c798b685f1fe9417e5987f#diff-98b4e0f87126f3486c8dda8f693ee45c682a856f17958aed5a33d51e5fd27766R887
https://github.com/PX4/PX4-Autopilot/commit/79c9af5a79c546f3b1c798b685f1fe9417e5987f#diff-d9bd9fba609cf44746b2a0a64e461dba64493d734810bad0d532873d9ec58020R120

File: src/modules/commander/state_machine_helper.cpp (hunk 4 above)
Comment (MaEtUgR) [reply src/modules/commander/state_machine_helper.cpp]: Adressed in https://github.com/PX4/PX4-Autopilot/pull/17785/commits/72543a94b56dd9a46cc2f3e871ef80bc250c5bf8

File: src/modules/commander/state_machine_helper.cpp (hunk 5 above)
Comment (MaEtUgR) [reply src/modules/commander/state_machine_helper.cpp]: It's only for loiter and takeoff. Mission finishes before returning when all links are lost.
I removed it now in https://github.com/PX4/PX4-Autopilot/pull/17785/commits/207b5705b94b4c95a0c94f1e9dafc543823c3f45 mainly because there could be SITL cases where you want to fly with `commander takeoff` and no link which would not be possible anymore afterwards. I don't see another real-world use case but maybe it exists. We can add the data link loss by default and then the user needs to explicitly disable both at which point I'd say he hopefully knows what he does.

**Comment by MaEtUgR** [comment]
Thanks for the review @julianoes ! I tried to address everything.

File: src/modules/commander/state_machine_helper.h (hunk 6 above)
Comment (julianoes) [reply src/modules/commander/state_machine_helper.h:118]: I see. Up to you.

**Review by julianoes — APPROVED** [review]

**Comment by Alkxndr** [comment]
@MaEtUgR I have been doing some testing with the RCL behavior to replicate BVLOS conditions before running my first full length BVLOS flight over 4G with a companion computer setup. The idea is to have me locally for manual control in case of any unexpected take-off/landing behavior and run the mission over 4g for the rest of the duration.

**Problem:**
When the RC loss is detected, telemetry links no longer have full control of aircraft behavior even if they are healthy. 

**Expected behavior:**
When RC is turned off or link deteriorates, control options via QGC including hold, continue, goto, altitude changes, etc. should still be available. 

**Actual behavior:**
When RC is turned off control options are limited. In all modes we can Land, or RTL, however entering hold or continuing mission is not always available depending on COM_RCL_EXCEPT settings. In hold mode it is not possible to continue the mission in QGC following RC link loss. In Offboard you cannot hold or continue the mission, only RTL and LAND are available. When attempting to select all 3 modes for COM_RCL_EXCEPT, only 2 can be chosen at a time. 

**Suggested fix:**
Switching of modes should be possible across all settings if link between vehicle and QGC is available. Only way to launch a mission and maintain full control via QGC is to start the mission without a RC connected at all which is obviously not acceptable.
